### PR TITLE
Fix/geolocation/integrate location service

### PR DESCRIPTION
**Pull request by lazarinibruno**
# Integrate Location Service and Remove old code

## Description
This PR integrates the newly written location service.  By doing so, the logic for the location is contained within the `LocationServiceImpl` class. Moreover, the app requests the access to the location at the right time (namely, when clicking on the Map). It closes issues #78 and #95. 

## Changes
- Removes the location access request logic from the main (since it is now handled by `LocationServiceImpl`).
- Instantiates a `locationService` in the main activity, which is then passed as an argument to `PeriodPalsApp` and consecutively to `MapScreen`.
- Adapts the `MapScreen` to the new implementation. 
- Adapted the `MapScreenTest`
 
#### Added
None.

#### Modified
- `MainActivity.kt`
- `Map.kt`
- `MapScreenTest.kt`

#### Removed
None.

## Dependencies Added
None.

## Testing
- I adapted the `MapScreenTest` to work with the new location service.
- I did two tests from the test that checks whether the map is shown when permission is granted:
    - One for when approximate location is granted
    - One for when precise location is granted 

## Screenshots
image

Location access request in the right screen (੭*ˊᵕˋ)੭

### CODE REVIEW

**Review by agonzalez-r — CHANGES_REQUESTED** [review]
## Summary
Thank you for this implementation! It contributes a lot on the Map feature of the app. 
Sadly I can't manually check if it works at expected (my emulator is only connected to wifi when I'm on campus for some reason so I can't login). But all the tests pass. 
## Important (which includes)
#### Code Quality
- I left a comment about the parameter `locationPermissionGranted` for `MapViewContainer()`, I think it should be replaced directly by `locationGrantedType` so you can directly use it 
#### Functionality
- ***Question***: where is the distinction between the precise and approximate location granted type treated? Since it's the same code for both types, I figure that it is automatically treated different by the location listener, but I'm not sure
## Testing
The tests for this screen are not complete: 
- Mainly, there is no tests to assert that components like `BottomNavigationMenu` and `TopAppBar` are correctly displayed (which also is not consistent with the rest of screen tests)
- For all tests (which difference is only the type of permission granted) you assert the same thing, only that map view exists. I think each tests should verify something more specific for each type (if possible). At least, when no permission is granted, could you verify that the default location is displayed? 


## Suggestions
- Tests more precise for different typer of granted permission types? (again, not sure if this is possible)
## Steps before PR approved
- Add tests for the correct display of `BottomNavigationMenu` and `TopAppBar` in `MapScreenTest.kt`
- Modify `MapViewContainer()` parameters for better code quality

**Comment by lazarinibruno** [comment]
> * For all tests (which difference is only the type of permission granted) you assert the same thing, only that map view exists. I think each tests should verify something more specific for each type (if possible). At least, when no permission is granted, could you verify that the default location is displayed?

Well, as @taghizadlaura said in some of the meetings, there is no direct way of testing the content of the map. This is why we only assert that the map is displayed given the different permissions. Moreover, this PR focuses on the integration of the location service, not in the map itself.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt [hunk 1]
```diff
@@ -33,13 +36,23 @@ private const val TAG = "MapView"
...
     navigationActions: NavigationActions
 ) {
   val context = LocalContext.current
   val fusedLocationClient = remember { LocationServices.getFusedLocationProviderClient(context) }
   val mapView = remember { MapView(context) }
 
+  locationService.requestUserPermissionForLocation()
+  val locationGrantedType = locationService.locationGrantedType.collectAsState().value
+
+  val locationPermissionGranted =
```
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/map/Map.kt:48]: I know. I did that as a temporary fix to integrate the location service. In S4, I will finish the location service to handle what to do given the different types of location access granted #130.

**Review by agonzalez-r — APPROVED** [review]
Okay, we'll get to the changes later.

File: app/src/main/java/com/android/periodpals/ui/map/Map.kt [hunk 2]
```diff
@@ -31,15 +34,21 @@ private val DEFAULT_LOCATION = GeoPoint(46.5191, 6.5668)
 private const val TAG = "MapView"
 
 @Composable
-fun MapScreen(
-    modifier: Modifier = Modifier,
-    locationPermissionGranted: Boolean,
-    navigationActions: NavigationActions
-) {
+fun MapScreen(locationService: LocationServiceImpl, navigationActions: NavigationActions) {
```
Comment (taghizadlaura) [inline app/src/main/java/com/android/periodpals/ui/map/Map.kt:37]: Problem with the declaration in the MainActivity

File: app/src/main/java/com/android/periodpals/MainActivity.kt [hunk 3]
```diff
@@ -143,7 +94,7 @@ fun PeriodPalsApp(
     // Map
     navigation(startDestination = Screen.MAP, route = Route.MAP) {
       composable(Screen.MAP) {
-        MapScreen(Modifier.fillMaxSize(), locationPermissionGranted, navigationActions)
+        MapScreen(Modifier.fillMaxSize(), locationService, navigationActions)
```
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/MainActivity.kt]: Should be fixed now.

File: app/src/androidTest/java/com/android/periodpals/ui/map/MapScreenTest.kt [hunk 4]
```diff
@@ -2,41 +2,131 @@ package com.android.periodpals.ui.map
...
+        mapOf(
+            Manifest.permission.ACCESS_FINE_LOCATION to true,
+            Manifest.permission.ACCESS_COARSE_LOCATION to true)
+
+    permissionCallbackCaptor.value.onActivityResult(permissions)
+
+    composeTestRule.setContent {
+      PeriodPalsAppTheme {
+        MapScreen(
+            modifier = Modifier,
```
Comment (taghizadlaura) [inline app/src/androidTest/java/com/android/periodpals/ui/map/MapScreenTest.kt:102]: Same problem as the previous one in the MainActivity

File: app/src/androidTest/java/com/android/periodpals/ui/map/MapScreenTest.kt [hunk 5]
```diff
@@ -2,41 +2,131 @@ package com.android.periodpals.ui.map
...
+            Manifest.permission.ACCESS_FINE_LOCATION to false,
+            Manifest.permission.ACCESS_COARSE_LOCATION to false)
+
+    permissionCallbackCaptor.value.onActivityResult(permissions)
+
     composeTestRule.setContent {
       PeriodPalsAppTheme {
         MapScreen(
-            locationPermissionGranted = false,
+            modifier = Modifier,
```
Comment (taghizadlaura) [inline app/src/androidTest/java/com/android/periodpals/ui/map/MapScreenTest.kt:123]: same here

File: app/src/androidTest/java/com/android/periodpals/ui/map/MapScreenTest.kt [hunk 6]
```diff
@@ -2,41 +2,131 @@ package com.android.periodpals.ui.map
...
+            Manifest.permission.ACCESS_COARSE_LOCATION to true)
+
+    // Simulate the permission grant
+    permissionCallbackCaptor.value.onActivityResult(permissions)
+
     composeTestRule.setContent {
       PeriodPalsAppTheme {
         MapScreen(
-            locationPermissionGranted = true,
+            modifier = Modifier,
```
Comment (taghizadlaura) [inline app/src/androidTest/java/com/android/periodpals/ui/map/MapScreenTest.kt:80]: same here

**Review by taghizadlaura — CHANGES_REQUESTED** [review]

**Review by taghizadlaura — APPROVED** [review]
LGTM now, but there is a bug on the approximate location. We decided to fix it in a different task : #130

**Pull request merged**